MOBILEAPPS-1952: Added External SSO flow to obtain an AcsApiClient

ahusseini opened the pull request.

ahusseini committed:
MOBILEAPPS-1952: Added new FSR classes to iOS Project

ahusseini committed:
MOBILEAPPS-1952: Completed external login flow

Commit:
Merge remote-tracking branch 'origin/develop' into MOBILEAPPS-1952

sleeke commented on acs-apiclient/AcsApiClient.cs:
```diff
@@ -119,8 +119,10 @@ private void InvokeLog(string str)
-            if (string.IsNullOrEmpty(config.PortalPassword) || string.IsNullOrEmpty(config.PortalUsername))
-            {
+            if (
+                !config.IsSSOClient 
```
:bowtie: The newline seems a little strange here

sleeke replied:
I think we need to come up with a convention... 🤔

sleeke replied:
✅

sleeke commented on acs-apiclient/AcsApiError.cs:
```diff
@@ -59,6 +59,27 @@ public enum AcsApiError
+        CouldNotExchangeToken,
+
+        /// <summary>The code provided by the Identity Provider is not valid.</summary>
+        InvalidCode,
```
💡 It might be worth being more specific to avoid future conflicts, e.g ```IdpCodeInvalid```

sleeke replied:
✅

sleeke commented on acs-apiclient/AcsApiError.cs:
```diff
@@ -59,6 +59,27 @@ public enum AcsApiError
+        InvalidCode,
+
+        /// <summary>The state returned by the Identity Provider does not match the one it received.</summary>
+        StateMismatch,
```
💡 Same here, maybe ```IdpStateMismatch```

sleeke replied:
✅

sleeke commented on acs-apiclient/AcsApiError.cs:
```diff
@@ -59,6 +59,27 @@ public enum AcsApiError
+        ServerError,
+
+        /// <summary>The login flow was interrupted by the user.</summary>
+        Interrupted,
```
💡 In the vein of being specific (as mentioned in the comments below), maybe this should be ```InterruptedByUser```?

sleeke replied:
✅

sleeke commented on acs-apiclient/AcsApiError.cs:
```diff
@@ -59,6 +59,27 @@ public enum AcsApiError
+        BadRequest,
+
+        /// <summary>The remote server returned a client error response.</summary>
+        ClientError,
```
❓ Is it worth differentiating between ClientError and BadRequest? Technically they're both Client Errors

sleeke replied:
✅

sleeke commented on acs-apiclient/ExternalFlowDelegate.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Used to inform the delegate that the user cancelled the login flow.
+        /// </summary>
+        void DidCancelLoginProcess();
```
💡 Again, maybe the more specific ```UserCancelledLogin``` would be better

sleeke replied:
✅

sleeke commented on acs-apiclient/ExternalFlowDelegate.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Used to inform the delegate that the callback was intercepted, and a code was successfully received.
+        /// </summary>
+        void DidRetrieveCodeSuccessfully();
```
💡 The use of ```DidDoSomething``` is very iOS-centric. Perhaps ```RetrievedCode``` would be better?

sleeke replied:
or maybe ```continueTokenExchange```

sleeke replied:
✅

sleeke commented on acs-apiclient/ExternalFlowDelegate.cs:
```diff
@@ -0,0 +1,34 @@
+        /// Used to inform the delegate of any errors encountered during the login flow.
+        /// </summary>
+        /// <param name="error">An AcsApiError describing the error that was encountered.</param>
+        void DidEncounterError(AcsApiError error);
```
💡 As above; maybe ```LoginFailedWithError```

sleeke replied:
...or perhaps something more specific if this isn't for the entire login process

sleeke replied:
✅

sleeke commented on acs-apiclient/LoginDelegate.cs:
```diff
@@ -0,0 +1,41 @@
+        /// Informs the delegate that the password flow should be used. This method is typically used for displaying 
+        /// the password field.
+        /// </summary>
+        void ShouldBeginPasswordFlow();
```
💡 I wonder if this should be called ```BeginNonSsoFlow```?

sleeke replied:
or ```BeginLocalFlow```?

sleeke commented on acs-apiclient/LoginController.cs:
```diff
@@ -0,0 +1,259 @@
+        /// Exchanges the intercepted callback code for an Access Token.
+        /// </summary>
+        /// <param name="code">The intercepted callback code.</param>
+        void BeginAccessTokenExchange(string code)
```
💡 Let's be specific about this; which token are we exchanging. To my mind, it could be:
 - OAuth1
 - OKTA

sleeke replied:
✅

sleeke commented on acs-apiclient/LoginController.cs:
```diff
@@ -0,0 +1,259 @@
+
+            acsApiClientConfiguration.IsSSOClient = true;
+            acsApiClientConfiguration.AccessToken = oauthToken;
+            acsApiClientConfiguration.AccessTokenSecret = oauthSecret;
```
💡 I think that now we're dealing with so many access tokens, keys and secrets, it would be worth being specific about which is which. Here, in the clientConfiguration, I think that the name OAuthToken and OAuthSecret would be better than AccessToken and AccessTokenSecret

ahusseini replied:
👍

sleeke replied:
✅

sleeke commented on acs-apiclient/LoginController.cs:
```diff
@@ -0,0 +1,259 @@
+            }
+            catch (Exception)
+            {
+                RunOnMainThread(() => loginDelegate.DidEncounterError(AcsApiError.CouldNotLogin));
```
❓ Could we add some more detail to this response? It seems like this could be for any number of reasons, and it'd be helpful to pass some info along...

sleeke replied:
I guess this is tricky since we only pass back an enum, but maybe it's worth passing more back? Let's chat about it

ahusseini replied:
👍

sleeke replied:
✅

sleeke commented on acs-apiclient/LoginController.cs:
```diff
@@ -0,0 +1,259 @@
+                    }
+                    throw new AcsApiException(AcsApiError.StateMismatch);
+                }
+                throw new AcsApiException(AcsApiError.InvalidCode);
```
❓ Is it worth differentiating between an empty state and an invalid code?

sleeke replied:
maybe this is more accurately an ```invalidRedirectParameters```?

ahusseini replied:
👍

sleeke replied:
✅

sleeke commented on acs-apiclient/Models/IdentityProvider.cs:
```diff
@@ -7,24 +10,46 @@ namespace AcsApi.Models
+                    from parameterValue in parameters.GetValues(parameterKey)
+                    select $"{ HttpUtility.UrlEncode(parameterKey) }={ HttpUtility.UrlEncode(parameterValue) }"
+                ).ToArray();
+                var parameterString = string.Join("&", encodedParameters);
```
❓ I wonder if it's worth pulling the formation of this string out to a utility method for readability's sake, if not for future convenience?

ahusseini replied:
👍

sleeke replied:
✅

sleeke commented on acs-apiclient/PlatformLoginController.cs:
```diff
@@ -0,0 +1,28 @@
+        /// <summary>
+        /// Platform specific implementation to launch and begin the external flow.
+        /// </summary>
+        /// <param name="context">The context from which the external flow will launch from.</param>
```
:bowtie: Should be "The context from which the external flow will launch"

ahusseini replied:
👍

sleeke replied:
✅

sleeke commented on acs-apiclient/PlatformLoginController.cs:
```diff
@@ -0,0 +1,28 @@
+        /// Runs an action on the main thread.
+        /// </summary>
+        /// <param name="action">The work to be run on the main thread, wrapped in an Action.</param>
+        protected abstract void RunOnMainThread(Action action);
```
❓ Should this be pulled out to a separate threading interface?

sleeke commented on acs-apiclient/iOS/ExternalFlowWebViewController.cs:
```diff
@@ -0,0 +1,168 @@
+        void PrepareBackButton()
+        {
+            var closeIcon = UIImage.FromBundle("ic_close_36pt");
+            dismissButton = new UIBarButtonItem(closeIcon, UIBarButtonItemStyle.Plain, (sender, e) =>
```
⚠️ The close and refresh icons don't look like they fit with each other. Since you've used the system's refresh button, perhaps we should also use the system's cancel (```UIBarButtonSystemItem.Stop```) button?

ahusseini replied:
Changed the refresh button to match the designs.

sleeke commented on acs-apiclient/iOS/ExternalFlowWebViewController.cs:
```diff
@@ -0,0 +1,168 @@
+            PrepareBackButton();
+            PrepareRefreshButton();
+            PrepareNavigationItem();
+            PrepareWebView();
```
👍 Nice tidy prepare function, I like the style 🙂

sleeke commented on acs-apiclient/iOS/LoginController.iOS.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            if (!(context is UIViewController))
+            {
+                throw new NotImplementedException(
```
:bowtie: Should be an ```ArgumentException```

ahusseini replied:
👍

sleeke replied:
✅

sleeke commented on acs-apiclient/iOS/ExternalFlowWebViewController.cs:
```diff
@@ -0,0 +1,168 @@
+            WKWebView webView,
+            WKNavigationAction navigationAction,
+            Action<WKNavigationActionPolicy> decisionHandler
+        ) {
```
:bowtie: We need to come unp with  convention here. I'm definitely more of a fan of keeping brackets on the same line as the final argument:
```
        public void DecidePolicy(
            WKWebView webView,
            WKNavigationAction navigationAction,
            Action<WKNavigationActionPolicy> decisionHandler) 
        {
            try
```

But also willing to bow to C# conventions or you guys' preference...

sleeke replied:
✅

sleeke reviewed (changes requested):
⚠️ This looks great in general, but there's a bunch of suggestions here for readability, clarity, etc., and some we should chat about another day 🙂

ahusseini committed:
MOBILEAPPS-1952: Fixed review issues

ahusseini committed:
Merge branch 'MOBILEAPPS-1952' of github.com:foreseecode/acs-apiclient-net into MOBILEAPPS-1952

sleeke commented on acs-apiclient/AcsApiClient.cs:
```diff
@@ -119,9 +119,8 @@ private void InvokeLog(string str)
         {
-            if (
-                !config.IsSSOClient 
-                && (string.IsNullOrEmpty(config.PortalPassword) || string.IsNullOrEmpty(config.PortalUsername))
```
❓ Look into the conventions here

sleeke commented on acs-apiclient/AcsApiClientConfig.cs:
```diff
@@ -87,14 +87,14 @@ public class AcsApiClientConfig
 
         /// <summary>
-        /// Gets or sets the Access token
+        /// A token received upon authenticating
```
💡 "A token received upon completing the entire login process. Used to access to the ForeSee services"

sleeke commented on acs-apiclient/AcsApiClientConfig.cs:
```diff
@@ -87,14 +87,14 @@ public class AcsApiClientConfig
 
         /// <summary>
-        /// Gets or sets the Access token secret
+        /// A secret received upon authenticating
```
💡 "A token received upon completing the entire login process. Used in conjunction with the OAuthToken"

sleeke commented on acs-apiclient/LoginController.cs:
```diff
@@ -117,49 +122,54 @@ void BeginOAuthTokenExchange(string accessToken)
+            catch (Exception exception)
             {
-                RunOnMainThread(() => loginDelegate.DidEncounterError(AcsApiError.CouldNotLogin));
+                RunOnMainThread(() => loginDelegate.EncounteredError(AcsApiError.Other, exception.Message));
```
❓ Where is the ```other``` type used?

sleeke commented on acs-apiclient/LoginController.cs:
```diff
@@ -217,43 +226,42 @@ public bool ShouldInterceptRequest(string urlString)
+            if (string.IsNullOrEmpty(authCode))
             {
-                loginDelegate.DidEncounterError(AcsApiError.InvalidCode);
+                loginDelegate.EncounteredError(AcsApiError.InvalidCallbackParameters, "The Auth Code is invalid.");
```
💡 "The retrieved Auth Code is invalid"

sleeke commented on acs-apiclient/LoginController.cs:
```diff
@@ -217,43 +226,42 @@ public bool ShouldInterceptRequest(string urlString)
         /// </summary>
         /// <param name="error">Error.</param>
-        public void DidEncounterError(AcsApiError error)
+        public void EncounteredError(AcsApiError error, string message, string code = "")
```
❓ Is there a reason this isn't passed as a number?

sleeke commented on acs-apiclient/iOS/ExternalFlowWebViewController.cs:
```diff
@@ -68,29 +68,31 @@ void PrepareView()
+        {
+            var refreshIcon = UIImage.FromBundle("ic_sync");
+            refreshButton = new UIBarButtonItem(refreshIcon, UIBarButtonItemStyle.Plain, (sender, e) => {
+				webView.LoadRequest(new NSUrlRequest(urlToLoad));            
```
⚠️ This seems to be the opposite of what I suggested. Check out how it looks

Commit:
MOBILEAPPS-1952: Review:
 - Switched to integer status codes
 - Updated docs
 - Patched project issues
 - Added userprefs to gitignore
